client/modules/IDE/components/Editor/index.jsx

```diff
@@ -65,564 +73,599 @@ import * as PreferencesActions from '../../actions/preferences';
-  constructor(props) {
-    super(props);
-    this.state = {
-      currentLine: 1
+  return {
+    toDOM() {
+      return dom;
+    },
+    ignoreEvent() {
+      return false; // Ensures that the widget doesn't block editor interaction
+    }
+  };
+};
+
+const Editor = (props) => {
+  const {
+    file,
+    fontSize,
+    unsavedChanges,
+    setUnsavedChanges,
+    updateFileContent,
+    autorefresh,
+    isPlaying,
+    clearConsole,
+    startSketch,
+    provideController,
+    isExpanded,
+    t,
+    collapseSidebar,
+    expandSidebar,
+    lintMessages,
+    clearLintMessage,
+    updateLintMessage,
+    lintWarning,
+    theme,
+    hideRuntimeErrorWarning,
+    autocompleteHinter
+  } = props;
+
+  const beepRef = useRef(null);
+
+  useEffect(() => {
+    beepRef.current = new Audio(beepUrl);
+
+    return () => {
+      beepRef.current = null;
     };
-    this._cm = null;
-    this.tidyCode = this.tidyCode.bind(this);
+  }, [beepUrl]);
 
-    this.updateLintingMessageAccessibility = debounce((annotations) => {
-      this.props.clearLintMessage();
+  const updateLintingMessageAccessibility = useCallback(
+    debounce((annotations) => {
+      clearLintMessage();
       annotations.forEach((x) => {
         if (x.from.line > -1) {
-          this.props.updateLintMessage(x.severity, x.from.line + 1, x.message);
+          updateLintMessage(x.severity, x.from.line + 1, x.message);
         }
       });
-      if (this.props.lintMessages.length > 0 && this.props.lintWarning) {
-        this.beep.play();
-      }
-    }, 2000);
-    this.showFind = this.showFind.bind(this);
-    this.showReplace = this.showReplace.bind(this);
-    this.getContent = this.getContent.bind(this);
-  }
-
-  componentDidMount() {
-    this.beep = new Audio(beepUrl);
-    // this.widgets = [];
-    this._cm = CodeMirror(this.codemirrorContainer, {
-      theme: `p5-${this.props.theme}`,
-      lineNumbers: this.props.lineNumbers,
-      styleActiveLine: true,
-      inputStyle: 'contenteditable',
-      lineWrapping: this.props.linewrap,
-      fixedGutter: false,
-      foldGutter: true,
-      foldOptions: { widget: '\u2026' },
-      gutters: ['CodeMirror-foldgutter', 'CodeMirror-lint-markers'],
-      keyMap: 'sublime',
-      highlightSelectionMatches: true, // highlight current search match
-      matchBrackets: true,
-      emmet: {
-        preview: ['html'],
-        markTagPairs: true,
-        autoRenameTags: true
-      },
-      autoCloseBrackets: this.props.autocloseBracketsQuotes,
-      styleSelectedText: true,
-      lint: {
-        onUpdateLinting: (annotations) => {
-          this.updateLintingMessageAccessibility(annotations);
-        },
-        options: {
-          asi: true,
-          eqeqeq: false,
-          '-W041': false,
-          esversion: 11
-        }
-      },
-      colorpicker: {
-        type: 'sketch',
-        mode: 'edit'
+      if (lintMessages.length > 0 && lintWarning) {
```

Comment:
You've got no dependencies in the array, so the values of the `lintMessages` and `lintWarning` variables will be what they were at the first render.  ie. a "stale closure".